ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -0,0 +1,309 @@
+  private val logger = ContextualizedLogger.get(getClass)
+  private[cache] val cacheOffset = new AtomicLong(0L)
+
+  def consumeFrom(implicit
+      loggingContext: LoggingContext
+  ): Flow[ContractStateEvent, Unit, NotUsed] =
+    Flow[ContractStateEvent]
+      .wireTap(debugEvents(_))
+      .alsoTo(Sink.foreach(updateCaches))
+      .map(updateOffsets)
+
+  override def lookupActiveContract(readers: Set[Party], contractId: ContractId)(implicit
+      loggingContext: LoggingContext
+  ): Future[Option[Contract]] =
+    contractsCache
+      .get(contractId)
+      .map(contractStateToResponse(_, readers, contractId))
+      .getOrElse(readThroughStateCache(contractId, readers))
+
+  override def lookupContractKey(readers: Set[Party], key: GlobalKey)(implicit
+      loggingContext: LoggingContext
+  ): Future[Option[ContractId]] =
+    keyCache
+      .get(key)
+      .map(keyStateToResponse(_, readers))
+      .map(Future.successful)
+      .getOrElse(readThroughKeyCache(key, readers))
+
+  override def lookupMaximumLedgerTime(ids: Set[ContractId])(implicit
+      loggingContext: LoggingContext
+  ): Future[Option[Instant]] =
+    if (ids.isEmpty)
+      Future.failed(EmptyContractIds())
+    else {
+      val (cached, toBeFetched) = partitionCached(ids)
+      if (toBeFetched.isEmpty)
+        Future.successful(Some(cached.max))
+      else
+        contractsReader
+          .lookupMaximumLedgerTime(toBeFetched)
+          .map(_.map(m => (cached + m).max))
+    }
+
+  private def partitionCached(
+      ids: Set[ContractId]
+  )(implicit loggingContext: LoggingContext) = {
+    val cacheQueried = ids.map(id => id -> contractsCache.get(id))
+    val cached = cacheQueried.collect {
+      case (_, Some(Active(_, _, createLedgerEffectiveTime))) => createLedgerEffectiveTime
+      case (_, Some(_)) => throw ContractNotFound(ids)
+    }
+    val missing = cacheQueried.collect { case (id, None) => id }
+    (cached, missing)
+  }
+
+  private def readThroughStateCache(contractId: ContractId, readers: Set[Party])(implicit
+      loggingContext: LoggingContext
+  ) = {
+    val currentCacheOffset = cacheOffset.get()
+    val fetchStateRequest =
+      Timed.future(
+        metrics.daml.index.lookupContract,
+        contractsReader.lookupContractState(contractId, currentCacheOffset),
+      )
+    val eventualValue = fetchStateRequest.map(toContractCacheValue)
+
+    for {
+      _ <- contractsCache.putAsync(
+        key = contractId,
+        validAt = currentCacheOffset,
+        eventualValue = eventualValue,
+      )
+      value <- eventualValue
+      response <- contractStateToResponse(value, readers, contractId)
+    } yield response
+  }
+
+  private def keyStateToResponse(
+      value: ContractKeyStateValue,
+      readers: Set[Party],
+  ): Option[ContractId] = value match {
+    case Assigned(contractId, createWitnesses) if nonEmptyIntersection(readers, createWitnesses) =>
+      Some(contractId)
+    case _ => Option.empty
```

Comment:
We probably already talked about default cases. This seems risky wrt future changes to the number of states a key can have. I'd therefore suggest to make the pattern matching complete to make the code easier to change.